FullStorySegmentMiddleware/Classes/FullStoryMiddleware.m

```diff
@@ -233,36 +226,59 @@ - (NSDictionary *) getDictionaryFromArrayObject:(NSArray *) arr withKey:(NSStrin
-    for (NSString *key in dict2) {
-        if (dict[key] != nil && [dict[key] isKindOfClass:[NSArray class]]) {
-            NSMutableArray *arr = [[NSMutableArray alloc] initWithArray:dict[key]];
-            [arr addObjectsFromArray:dict2[key]];
-            [dict removeObjectForKey:key];
-            [dict setObject:arr forKey:key];
-        } else if (dict[key] != nil) {
-            // if the current value is not an array already, then delete it and upgrade it to array
-            NSArray *arr = [[NSArray alloc]initWithObjects:dict[key], dict2[key], nil];
-            [dict removeObjectForKey:key];
-            [dict setObject:arr forKey:[key stringByAppendingString:@"s"]];
+    // when adding a parsed dict into the result dict, emurate and add each object
+    for (NSString *k in dict2) {
+        NSString *nestedKey;
+        if ([key length] > 0) {
+            nestedKey = [[key stringByAppendingString:@"."] stringByAppendingString:k];
         } else {
-             [dict setObject:dict2[key] forKey:key];
+            nestedKey = k;
         }
+        [self appendToDictionary:dict withKey:nestedKey andValue:dict2[k]];
     }
 }
 
--(void) appendToDictionary:(NSMutableDictionary *) dict withKey:(NSString *) key andValue:(NSObject *) obj {
-    // when adding a parsed array into the result dict, check if the key with suffix already exsist, if so then we need to append to the result arrays insead of replacing the object.
-    if (dict[key] != nil && [dict[key] isKindOfClass:[NSArray class]]) {
-        [dict[key] addObject:obj];
-    } else if (dict[key] != nil) {
-        // if the current value is not an array already, then delete it and upgrade it to array
-        NSArray *arr = [[NSArray alloc]initWithObjects:dict[key], obj, nil];
-        NSString *suffixedKey = [key stringByAppendingString:@"s"];
+-(void) appendToDictionary:(NSMutableDictionary *) dict withKey:(NSString *) key andArray:(NSArray *) arr {
+    // when adding a parsed array into the result dict, emurate and add each object
+    for (id obj in (NSArray *)arr) {
+        [self appendToDictionary:dict withKey:key andValue:obj];
+    }
+}
+
+-(void) appendToDictionary:(NSMutableDictionary *) dict withKey:(NSString *) key andSimpleObject:(NSObject *) obj {
+    // obj is simple, add it into the result dict, check if the key with suffix already exsist, if so then we need to append to the result arrays insead of replacing the object.
+    // this creates a mutable array each time an object is added, maybe we should consider making arrays mutable in the dict
+    // key is already suffixed, we just need to check if it ends with 's' to know if it's plural
+    bool isPlural = [key hasSuffix:@"s"];
+    NSString *pluralKey = [key stringByAppendingString:@"s"];
+    NSString *singularKey = [key substringToIndex:[key length] - 1];
```

Comment:
I'd be inclined to put the definition of pluralKey inside the `if (isPlural)` and put the definition of `singularKey` inside the `else`

Reply:
Also, I think that if key happened to be @"" then calling substringToIndex with a value of -1 will cause an exception.